File: voiture-resevation-backend/src/main/java/com/example/voitureresevation/entity/validators/VoitureValidator.java
package com.example.voitureresevation.entity.validators;

import com.example.voitureresevation.entity.Voiture;
import com.example.voitureresevation.zutils.validators.Validator;
import com.example.voitureresevation.zutils.validators.ValidatorItem;

import java.util.List;

public class VoitureValidator extends Validator<Voiture> {
    public VoitureValidator(Voiture item) {
        super(item);
    }

    public static void validate(Voiture item) {
        new VoitureValidator(item).validate();
    }

    public ValidatorItem<String> marque = new ValidatorItem<>(
            "marque",
            () -> this.getItem().getMarque(),
            (String value) -> {
                this.marque.getValidators()
                        .required()
                        .valid();
            }
    );
    public ValidatorItem<String> coleur = new ValidatorItem<>(
            "coleur",
            () -> this.getItem().getColeur(),
            (String value) -> {
                this.coleur.getValidators()
                        .required()
                        .valid();
            }
    );
    public ValidatorItem<String> categorie = new ValidatorItem<>(
            "categorie",
            () -> this.getItem().getCategorie(),
            (String value) -> {
                this.categorie.getValidators()
                        .required()
                        .valid();
            }
    );
    public ValidatorItem<Integer> nbrPlaces = new ValidatorItem<>(
            "nbrPlaces",
            () -> this.getItem().getNbrPlaces(),
            (Integer value) -> {
                this.nbrPlaces.getValidators()
                        .required()
                        .valid();
            }
    );
    public ValidatorItem<Integer> vitesse = new ValidatorItem<>(
            "vitesse",
            () -> this.getItem().getVitesse(),
            (Integer value) -> {
                this.vitesse.getValidators()
                        .required()
                        .valid();
            }
    );

    @Override
    public List<ValidatorItem<?>> validatorItems() {
        return List.of(
                this.marque,
                this.coleur,
                this.categorie,
                this.nbrPlaces,
                this.vitesse
        );
    }
}

File: voiture-resevation-frontend/src/app/controller/validators/voiture.validator.ts
import {Validator, ValidatorItem} from "@bshg/validation";
import {Voiture} from "src/app/controller/entities/voiture";

export class VoitureValidator extends Validator<Voiture> {
  marque = new ValidatorItem<string>(
    () => this.item().marque,
    (value) => this.item().marque = value,
    (value) => {
      this.marque.stringValidators
        ?.required()
        ?.valid()
    }
  )
  coleur = new ValidatorItem<string>(
    () => this.item().coleur,
    (value) => this.item().coleur = value,
    (value) => {
      this.coleur.stringValidators
        ?.required()
        ?.valid()
    }
  )
  categorie = new ValidatorItem<string>(
    () => this.item().categorie,
    (value) => this.item().categorie = value,
    (value) => {
      this.categorie.stringValidators
        ?.required()
        ?.valid()
    }
  )
  nbrPlaces = new ValidatorItem<number>(
    () => this.item().nbrPlaces,
    (value) => this.item().nbrPlaces = value,
    (value) => {
      this.nbrPlaces.numberValidators
        ?.required()
        ?.valid()
    }
  )
  vitesse = new ValidatorItem<number>(
    () => this.item().vitesse,
    (value) => this.item().vitesse = value,
    (value) => {
      this.vitesse.numberValidators
        ?.required()
        ?.valid()
    }
  )

  override validatorItems: ValidatorItem<any>[] = [
    this.marque,
    this.coleur,
    this.categorie,
    this.nbrPlaces,
    this.vitesse,
  ]

  getValidatorItem(fieldName: string): ValidatorItem<any> {
    // @ts-ignore
    return this[fieldName];
  }

  static init(item: () => Voiture): VoitureValidator {
    const validator = new VoitureValidator()
    validator.item = item
    return validator;
  }

}
